File: dotnet/samples/SimpleOrm.Sample/Models/Views/UserTransactionTotal.cs
namespace SimpleOrm.Sample.Models;

/// <summary>
/// View <c>user_transaction_totals</c>: one row per user with transaction totals.
/// Self-contained (ADR-0008 addendum 3): the defining SELECT lives in the attribute
/// and <c>db.CreateViewAsync</c> generates the CREATE VIEW. Read-only; declares
/// <c>[Key]</c> so read-by-key works at milestone 7, but never <c>[Generated]</c> —
/// nothing is written to a view. Not a <c>BaseModel</c>: projections carry no audit
/// columns.
/// </summary>
[View("user_transaction_totals", """
    select u.id              as user_id,
           u.name            as user_name,
           count(t.id)       as transaction_count,
           coalesce(sum(t.amount), 0) as total_amount
    from users u
    left join transactions t on t.user_id = u.id
    group by u.id, u.name
    """)]
public sealed class UserTransactionTotal
{
    [Key]
    [Column]
    public long UserId { get; set; }

    [Column]
    public required string UserName { get; set; }

    [Column]
    public int TransactionCount { get; set; }

    [Column]
    public decimal TotalAmount { get; set; }
}

File: php/samples/SimpleOrm.Sample/src/Models/Views/UserTransactionTotal.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Sample\Models\Views;

use DateTimeImmutable;
use SimpleOrm\Metadata\Attributes\Column;
use SimpleOrm\Metadata\Attributes\Key;
use SimpleOrm\Metadata\Attributes\View;
use SimpleOrm\Metadata\ColumnType;
use SimpleOrm\Types\Decimal;

/**
 * View `user_transaction_totals`: one row per user with transaction totals.
 * Self-contained (ADR-0008 add.3): the defining SELECT lives in the attribute
 * and `Db::createView()` / the view migration generate the CREATE VIEW.
 * Read-only; declares `#[Key]` so read-by-key works, but never `#[Generated]`:
 * nothing is written to a view. Not a `BaseModel`: projections carry no audit
 * columns.
 */
#[View('user_transaction_totals', <<<'SQL'
    select u.id              as user_id,
           u.name            as user_name,
           count(t.id)       as transaction_count,
           coalesce(sum(t.amount), 0) as total_amount,
           max(t.created_at) as last_transaction_at
    from users u
    left join transactions t on t.user_id = u.id
    group by u.id, u.name
    SQL)]
final class UserTransactionTotal
{
    #[Key]
    #[Column]
    public int $userId;

    #[Column]
    public string $userName;

    #[Column(type: ColumnType::Int32)]
    public int $transactionCount;

    #[Column]
    public Decimal $totalAmount;

    /** Added by migration V0006; null for users with no transactions. */
    #[Column('last_transaction_at')]
    public ?DateTimeImmutable $lastTransactionAtUtc = null;
}
